rules-java-diva/addon/src/main/java/org/jboss/windup/rules/apps/diva/service/GetOrCreateGraphService.java

```diff
@@ -0,0 +1,45 @@
+package org.jboss.windup.rules.apps.diva.service;
+
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.jboss.windup.graph.GraphContext;
+import org.jboss.windup.graph.model.WindupFrame;
+import org.jboss.windup.graph.model.WindupVertexFrame;
+import org.jboss.windup.graph.service.GraphService;
+
+public class GetOrCreateGraphService<T extends WindupVertexFrame> extends GraphService<T> {
+
+    public String TYPE;
+
+    public GetOrCreateGraphService(GraphContext context, Class<T> type) {
+        super(context, type);
+        try {
+            TYPE = (String) type.getField("TYPE").get(null);
+        } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public T getOrCreate(String key, String value, String... kvs) {
```

Comment:
It seems like this method is searching based on a set of key-value properties which is pretty close to what [`findAllByProperties`](https://github.com/windup/windup/blob/fca676af560e471fcd6631a2499bdf6cd7c596da/graph/api/src/main/java/org/jboss/windup/graph/service/GraphService.java#L83) method does.
I think we should try to leverage all the existing code base to reuse the code and, in case something is not implemented yet, contribute to the `GraphService` itself to improve just once and then let all the other components to leverage the improvements.
I feel like this class should be removed to leverage what `GraphService` or improving it.

Reply:
Thanks.  I will check GraphService code and consider your suggestion.